core/trie/proof.go

```diff
@@ -54,623 +55,586 @@ func (e *Edge) Len() uint8 {
+			}
+			curPos++
+		case *Edge: // Edge nodes represent paths between binary nodes
+			if !verifyEdgePath(&key, node.Path, curPos) {
+				return &felt.Zero, nil
+			}
 
-		if leftExist && rightExist {
-			return
-		} else if leftExist {
-			traverseNodes(nodeLeft, path, nodeHashes)
-		} else if rightExist {
-			traverseNodes(nodeRight, path, nodeHashes)
+			// Move to the immediate child node
+			curPos += node.Path.Len()
+			expectedHash = node.Child
 		}
-	case *Edge:
-		edgeNode, exist := nodeHashes[*currNode.Child]
-		if exist {
-			traverseNodes(edgeNode, path, nodeHashes)
+
+		// We've consumed all bits in our path
+		if curPos >= keyLen {
+			return expectedHash, nil
 		}
 	}
 }
 
-// MergeProofPaths removes duplicates and merges proof paths into a single path
-// merges paths in the specified order [commonNodes..., leftNodes..., rightNodes...]
-// ordering of the merged path is not important
-// since SplitProofPath can discover the left and right paths using the merged path and the rootHash
-func MergeProofPaths(leftPath, rightPath []ProofNode, hash hashFunc) ([]ProofNode, *felt.Felt, error) {
-	merged := []ProofNode{}
-	minLen := min(len(leftPath), len(rightPath))
-
-	if len(leftPath) == 0 || len(rightPath) == 0 {
-		return merged, nil, errors.New("empty proof paths")
-	}
-
-	if !leftPath[0].Hash(hash).Equal(rightPath[0].Hash(hash)) {
-		return merged, nil, errors.New("roots of the proof paths are different")
+// VerifyRangeProof checks whether the given key-value pairs and range proof are valid for the given root hash.
+// The range of key-value pairs should be consecutive (no gap inside) and monotonic increasing.
+// The given range proof contains 2 edge proofs. One is for the first key and the other is for the last key.
+// Both of them can be existent or non-existent.
+// This function also handles the following special cases:
+//
+//   - All elements proof: In this case the proof can be nil, but the range should be all the leaves in the trie.
+//   - One element proof: In this case, both left and right edge proofs are the same and the range should contain only one element.
+//   - Zero element proof: In this case a single edge proof is enough to verify. If there are still more elements, the proof is invalid.
+//
+// If the range proof is invalid, this function returns false and an error.
+// Otherwise, it will return a flag to indicate if there are more elements.
+//
+// TODO(weiihann): Given a binary leaf and a left-sibling first key, if the right sibling is removed, the proof would still be valid.
+// Conversely, given a binary leaf and a right-sibling last key, if the left sibling is removed, the proof would still be valid.
+// Range proof should not be valid for both of these cases, but currently is, which is an attack vector.
+// The problem probably lies in how we do root hash calculation.
+func VerifyRangeProof(root, first *felt.Felt, keys, values []*felt.Felt, proof *ProofNodeSet) (bool, error) { //nolint:funlen,gocyclo
```

Comment:
Why we need `first` here - which is if I'm correct `keys[0]`?
I noticed it is used in a special empty keys case - which is very weird imo - what's the purpose to proving empty range?

Reply:
In snap sync, a client provide the start key (here I use `first`) and the number of leaves for the range. In the returned keys, `first` may or may not equal to `key[0]`. If it's equal, then it's an existent edge proof, otherwise non-existent.